mirrord/layer/src/lib.rs

```diff
@@ -473,18 +458,93 @@ fn layer_start(config: LayerConfig) {
     LISTEN_PORTS
         .set(config.feature.network.incoming.listen_ports.clone())
         .expect("Setting LISTEN_PORTS failed");
+}
+
+/// Occurs after [`layer_pre_initialization`] has succeeded.
+///
+/// Starts the main parts of mirrord-layer.
+///
+/// ## Details
+///
+/// Sets up a few things based on the [`LayerConfig`] given by the user:
+///
+/// 1. [`tracing_subscriber`], or [`mirrord_console`];
+///
+/// 2. Connects to the mirrord-agent with [`connection::connect`];
+///
+/// 3. Initializes some of our globals;
+///
+/// 4. Replaces the [`libc`] calls with our hooks with [`enable_hooks`];
+///
+/// 5. Starts the main mirrord-layer thread.
+fn layer_start(config: LayerConfig) {
+    // does not need to be atomic because on the first call there are never other threads.
+    // Will be false when manually called from fork hook.
+    if LAYER_INITIALIZED.get().is_none() {
+        // If we're here it's not a fork, we're in the ctor.
+        let _ = LAYER_INITIALIZED.set(());
+        init_tracing(&config);
+        set_globals(&config);
+        enable_hooks(
+            config.feature.fs.is_active(),
+            config.feature.network.dns,
+            config
+                .sip_binaries
+                .clone()
+                .map(|x| x.to_vec())
+                .unwrap_or_default(),
+        );
+    }
 
-    enable_hooks(
-        file_mode.is_active(),
-        config.feature.network.dns,
-        config
-            .sip_binaries
-            .clone()
-            .map(|x| x.to_vec())
+    let _detour_guard = DetourGuard::new();
+    info!("Initializing mirrord-layer!");
+    trace!(
+        "Loaded into executable: {}, on pid {}, with args: {:?}",
+        std::env::current_exe()
+            .map(|path| path.to_string_lossy().to_string())
             .unwrap_or_default(),
+        std::process::id(),
+        std::env::args()
     );
 
-    RUNTIME.block_on(start_layer_thread(tx, rx, receiver, config));
+    let address = config
+        .connect_tcp
+        .as_ref()
+        .expect("layer loaded without proxy address to connect to")
+        .parse()
+        .expect("couldn't parse proxy address");
+
+    // SAFETY: This function runs once per process, `RUNTIME` is not used anywhere else but this
+    // function, so there are no other threads using `RUNTIME`, so it's safe to mutate it here.
+    let new_runtime = unsafe {
+        // leak the old runtime if there is one (on fork there is).
+        if let Some(old_runtime) = RUNTIME.take() {
+            mem::forget(old_runtime);
```

Comment:
given we leak the old runtime anyway always, I think maybe the solution should be stop having RUNTIME as a global and just leak it as well, always.

Reply:
forgotten means none will drop it, don't think it's bad but maybe worth adding as a comment as an option

Reply:
https://github.com/metalbear-co/mirrord/pull/1591/commits/db6a1c710909007f3cd84ab6163d059949e02ec6 added comment